ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm

```diff
@@ -97,6 +97,118 @@ static void stretchChildrenAlongCrossDimension(std::vector<CKStackUnpositionedIt
   }
 }
 
+/** The threshold that determines if a violation has actually occurred. */
+static const CGFloat kViolationEpsilon = 0.01;
+
+/**
+ Returns a lambda that computes the relevant flex factor based on the given violation.
+ @param violation The amount that the stack layout violates its size range.  See header for sign interpretation.
+ */
+static std::function<CGFloat(const CKStackUnpositionedItem &)> flexFactorInViolationDirection(const CGFloat violation)
+{
+  if (fabs(violation) < kViolationEpsilon) {
+    return [](const CKStackUnpositionedItem &item) { return 0; };
+  } else if (violation > 0) {
+    return [](const CKStackUnpositionedItem &item) { return item.child.flexGrow; };
+  } else {
+    return [](const CKStackUnpositionedItem &item) { return item.child.flexShrink; };
+  }
+}
+
+static inline CGFloat scaledFlexShrinkFactor(const CKStackUnpositionedItem &item, const CKStackLayoutComponentStyle &style)
+{
+  return stackDimension(style.direction, item.layout.size) * item.child.flexShrink;
+}
+
+/**
+ Returns a lambda that computes a flex shrink adjustment for a given item based on the provided violation.
+ @param items The unpositioned items from the original unconstrained layout pass.
+ @param style The layout style to be applied to all children.
+ @param violation The amount that the stack layout violates its size range.
+ @return A lambda capable of computing the flex shrink adjustment, if any, for a particular item.
+ */
+static std::function<CGFloat(const CKStackUnpositionedItem &, BOOL)> flexShrinkAdjustment(const std::vector<CKStackUnpositionedItem> &items,
+                                                                                          const CKStackLayoutComponentStyle &style,
+                                                                                          const CGFloat violation)
+{
+  const CGFloat scaledFlexShrinkFactorSum = std::accumulate(items.begin(), items.end(), 0, [&](CGFloat x, const CKStackUnpositionedItem &item) {
+    return x + scaledFlexShrinkFactor(item, style);
+  });
+  return [style, scaledFlexShrinkFactorSum, violation](const CKStackUnpositionedItem &item, BOOL isFirstFlex) {
+    const CGFloat scaledFlexShrinkFactorRatio = scaledFlexShrinkFactor(item, style) / scaledFlexShrinkFactorSum;
+    // The item should shrink proportionally to the scaled flex shrink factor ratio computed above.
+    // Unlike the flex grow adjustment the flex shrink adjustment needs to take the size of each item into account.
+    return -fabs(scaledFlexShrinkFactorRatio * violation);
```

Comment:
Unlike the flex grow adjustment below the flex shrink adjustment doesn't allocate the remaining violation to the first flexible item. This does mean that the resulting layout could result in pixel misalignment. You can see this behavior in the `testOverflowBehaviors_flex@2x.png` reference image below.